driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java

```diff
@@ -633,19 +632,34 @@ private <T> T getCommandResult(final Decoder<T> decoder, final ResponseBuffers r
     @Override
     public void sendMessage(final List<ByteBuf> byteBuffers, final int lastRequestId) {
         notNull("stream is open", stream);
-
         if (isClosed()) {
             throw new MongoSocketClosedException("Cannot write to a closed stream", getServerAddress());
         }
-
         try {
             stream.write(byteBuffers);
         } catch (Exception e) {
             close();
-            throw translateWriteException(e);
+            throwTranslatedWriteException(e);
         }
     }
 
+    @Override
+    public void sendMessageAsync(final List<ByteBuf> byteBuffers, final int lastRequestId,
+            final SingleResultCallback<Void> callback) {
+        beginAsync().thenRun((c) -> {
+            notNull("stream is open", stream);
+            if (isClosed()) {
+                throw new MongoSocketClosedException("Cannot write to a closed stream", getServerAddress());
+            }
+            c.complete(c);
```

Comment:
The purpose is actually the opposite, to prevent misuse of the void `complete()` when a result should be returned. This is a workaround to force a typecheck on the return type. It is very easy to make mistakes in the boilerplate (calling the wrong callback, passing the wrong value, and so on), whether in the old or new async code. I think it is important to prevent this (which is a correctness issue, and constant problem), even if it makes the API unusual (which is only ugly but not a correctness issue, and we only have to pay the price of getting used to this a few times).